src/Microsoft.AspNet.Mvc.TagHelpers/Internal/GlobbingUrlBuilder.cs

```diff
@@ -73,93 +80,105 @@ public GlobbingUrlBuilder(IFileProvider fileProvider, IMemoryCache cache, PathSt
         /// <param name="includePattern">The file globbing include pattern.</param>
         /// <param name="excludePattern">The file globbing exclude pattern.</param>
         /// <returns>The list of URLs</returns>
-        public virtual ICollection<string> BuildUrlList(string staticUrl, string includePattern, string excludePattern)
+        public virtual IReadOnlyList<string> BuildUrlList(
+            string staticUrl,
+            string includePattern,
+            string excludePattern)
         {
-            var urls = new HashSet<string>(StringComparer.Ordinal);
+            // Get urls that match the globbing patterns specified
+            var globbedUrls = ExpandGlobbedUrl(includePattern, excludePattern);
 
-            // Add the statically declared url if present
-            if (staticUrl != null)
+            if (staticUrl == null)
             {
-                urls.Add(staticUrl);
+                return globbedUrls;
             }
 
-            // Add urls that match the globbing patterns specified
-            var matchedUrls = ExpandGlobbedUrl(includePattern, excludePattern);
-            urls.UnionWith(matchedUrls);
+            var urls = new List<string>(1 + globbedUrls.Count)
+            {
+                staticUrl
+            };
+
+            string lastAddedEntry = null;
+            for (var i = 0; i < globbedUrls.Count; i++)
+            {
+                if (!string.Equals(staticUrl, globbedUrls[i], StringComparison.Ordinal))
+                {
+                    lastAddedEntry = globbedUrls[i];
+                    urls.Add(lastAddedEntry);
+                }
+            }
 
             return urls;
         }
 
-        private IEnumerable<string> ExpandGlobbedUrl(string include, string exclude)
+        private IReadOnlyList<string> ExpandGlobbedUrl(string include, string exclude)
         {
             if (string.IsNullOrEmpty(include))
             {
-                return Enumerable.Empty<string>();
+                return EmptyList;
             }
 
-            var includePatterns = include.Split(PatternSeparator, StringSplitOptions.RemoveEmptyEntries);
-            var excludePatterns = exclude?.Split(PatternSeparator, StringSplitOptions.RemoveEmptyEntries);
-
-            if (includePatterns.Length == 0)
-            {
-                return Enumerable.Empty<string>();
-            }
-
-            if (Cache != null)
+            var cacheKey = new GlobbingUrlKey(include, exclude);
```

Comment:
@rynowak the cache identifier does not capture the values of `include` \ `exclude` which is what we want to unique-ify on.